File: visualbasic/fib.vb
imports System.Math
' Imports BigInteger = System.Numerics.BigInteger

module Fibonacci
  function Fib(ByVal n as Integer) as Decimal
    if n < 2 then return n
    return Fib(n - 1) + Fib(n - 2)
  end function

  function FibFormula(ByVal n as Integer) as Decimal
    return Math.Round(Math.Pow((Math.Pow(5, .5) + 1) / 2, n) / Math.Pow(5, .5))
  end function

  ' Want to use BigInteger here
  function FibLinear(ByVal n as Integer) as Integer
    if n < 2 then return n
    dim prevFib as Integer = 0
    dim fib as Integer = 1
    while n > 1
      dim temp as Integer = fib
      fib = prevFib + fib
      prevFib = temp
      n -= 1
    end while
    return fib
  end function

  function FibTailRecursive(
      ByVal n as Integer,
      Optional ByVal prevFib as Integer = 0,
      Optional ByVal fib as Integer = 1
    )
    if n = 0 then return prevFib
    return FibTailRecursive(n - 1, fib, prevFib + fib)
  end function

  sub Main(ByVal cmdArgs() As String)
    dim input as Integer
    if cmdArgs.Length > 0 then
      input = cmdArgs(0)
    else
      input = 29
    end if
    Console.WriteLine(FibLinear(input))
  end sub
end module

File: language/csharp/fib/fib.cs
﻿static decimal Fib(int n)
{
  return n < 2 ? n : Fib(n - 1) + Fib(n - 2);
}

static decimal FibLinear(int n, decimal prevFib = 0, decimal fib = 1)
{
  if (n < 2) return n;
  while (--n > 0)
  {
    var temp = prevFib + fib;
    prevFib = fib;
    fib = temp;
  }
  return fib;
}

static decimal FibFormula(int n)
{
  return (decimal)Math.Round(Math.Pow((Math.Pow(5, .5) + 1) / 2, n) / Math.Pow(5, .5));
}

static decimal FibTailRecursive(int n, decimal prevFib = 0, decimal fib = 1)
{
  return n == 0 ? prevFib : FibTailRecursive(n - 1, fib, prevFib + fib);
}

static Tuple<decimal, decimal> FibFastAuxiliary(decimal n)
{
  if (n == 0) return new Tuple<decimal, decimal>(0, 1); // BigInteger.Zero, BigInteger.One;
  var fibs = FibFastAuxiliary(Math.Floor(n / 2));
  var a = fibs.Item1 * (fibs.Item2 * 2 - fibs.Item1);
  var b = fibs.Item1 * fibs.Item1 + fibs.Item2 * fibs.Item2;
  return n % 2 == 0
    ? new Tuple<decimal, decimal>(a, b)
    : new Tuple<decimal, decimal>(b, a + b);
}

static decimal FibFastDoubling(int n)
{
  return FibFastAuxiliary((decimal)n).Item1;
}

static IEnumerable<decimal> FibGenerator(int n, decimal prevFib = 0, decimal fib = 1)
{
  while (n-- >= 0)
  {
    yield return prevFib;
    var temp = prevFib + fib;
    prevFib = fib;
    fib = temp;
  }
}

static void FibGenerate(int n)
{
  foreach (decimal x in FibGenerator(n))
  {
    Console.WriteLine(x);
  }
}

var input = args.Length > 0 ? int.Parse(args[0]) : 29;
var result = FibFastDoubling(input);
Console.WriteLine(String.Format("Fibonacci {0}: {1}", input, result));
